pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,353 @@
+
+		if parentDir == "" || parentDir == "/" {
+			break
+		}
+
+		currentdir = path.Clean(parentDir)
+	}
+
+	slices.Reverse(dirs)
+
+	return dirs
+}
+
+// MkdirAll creates directories recursively on the FTP server. Here, os.FileMode is unused.
+// Here, os.FileMode is unused, but is added to comply with FileSystem interface.
+func (f *fileSystem) MkdirAll(name string, _ os.FileMode) error {
+	defer f.processLog(&FileLog{Operation: "MkdirAll", Location: path.Join(f.config.RemoteDir, name)}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("MkdirAll failed. Provide a valid path : %v", errEmptyPath)
+		return errEmptyPath
+	}
+	// returns a slice of those directories that do not exist with the first index being the latest existing parent directory path.
+	dirs := f.mkdirAllHelper(name)
+
+	currentDir := dirs[0]
+
+	for i, dir := range dirs {
+		if i == 0 {
+			continue
+		}
+
+		currentDir = path.Join(currentDir, dir)
+
+		err := f.conn.MakeDir(currentDir)
+		if err != nil {
+			f.logger.Errorf("MkdirAll failed : %v", err)
+
+			return err
+		}
+	}
+
+	f.logger.Logf("Directories creation completed successfully.")
+
+	return nil
+}
+
+// Open retrieves a file from the FTP server and returns a file handle.
+func (f *fileSystem) Open(name string) (container.File, error) {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Open", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Open_file failed. Provide a valid filename : %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		f.logger.Errorf("Open_file failed. Error opening file : %v", err)
+		return nil, err
+	}
+
+	defer res.Close()
+
+	filename := path.Base(filePath)
+
+	f.logger.Logf("Open_file successful. Filepath : %q", filePath)
+
+	return &file{
+		response: res,
+		name:     filename,
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+		metrics:  f.metrics,
+	}, nil
+}
+
+// permissions are not clear for Ftp as file commands do not accept an argument and don't store their file permissions.
+// currently, this function just calls the Open function.
+// Here, os.FileMode is unused, but is added to comply with FileSystem interface.
+func (f *fileSystem) OpenFile(name string, _ int, _ os.FileMode) (container.File, error) {
+	return f.Open(name)
+}
+
+// Remove deletes a file from the FTP server.
+func (f *fileSystem) Remove(name string) error {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Remove", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Remove_file failed. Provide a valid filename : %v", errEmptyFilename)
+		return errEmptyFilename
+	}
+
+	err := f.conn.Delete(filePath)
+
+	var textprotoError *textproto.Error
+	switch {
+	case errors.As(err, &textprotoError) && textprotoError.Code == ftp.StatusClosingDataConnection && textprotoError.Msg == "Transfer complete.":
```

Comment:
Thanks about doing the change.

I'm now thibk about something that could help I think.

I let you check.

If I understand well the issue this condition is trying to resolve.

- An error can be returned when deleting a file, even if the file was deleted anyway. It might depend on server implementation.

- when this occurs we receive the code

Reply:
But, I'm guessing the text message may vary with FTP server.

So I'm thinking about this.

We are lucky because this strange behavior is only happening when we are deleting a file.

So we could check for the file presence after we receive the code.

This way, no need to check the error message

We receive the code, OK, but does the file exist? No, then it worked. Otherwise something unexpected occurred

Reply:
I noticed this behaviour sometimes...I guess the behaviour was observed because of response not being closed somewhere before calling the remove function. Because I am not able to replicate the behaviour anymore. I feel we can do without this check.

Reply:
That would be great in fact.

Also, if someone faces it. Then they will receive the error linked to their FTP server

So, maybe we could remove and add a comment to the method saying we noticed some server may return an error even if the delete is successful